### DISCUSSION TITLE
Porting a arduino i2s configuration to devicescript

### DISCUSSION THREAD

**Discussion by cskiwi** [microsoft/devicescript#677] [Q&A]
Hi, 

For my girlfriends PhD i'm working on some sound loggers that can be placed next to a birds nest, these loggers record the decibels every second and record 2 min of sound every 10 min (so 8 min without any recording)

I started with a NodeMCU (ESP32) board and programmed some arduino code to do this.
This all worked mostly but I have some weird memory issues hat I can't figure out on how to resolve this due to my lack of C++ programming and I found the arduino way of working very developer friendly.

I'm a typescript (angular) developer. So once I found `devicescript` was of course very happy that I might just port the project to typescript and work with a familiar programming language.

As all loggers are already soldered and finished on the hardware side, I'd like to continue from this. So now comes my first challenge.
Getting the mic working in `devicescript`:

In `arduino` the config is as following:

```cpp
#define I2S_WS GPIO_NUM_22
#define I2S_SD GPIO_NUM_21
#define I2S_SCK GPIO_NUM_26
#define I2S_PORT_NUM I2S_NUM_0
#define I2S_SAMPLE_RATE (44100)
#define I2S_READ_LEN (I2S_SAMPLE_RATE * I2S_NUM_CHANNELS * (I2S_SAMPLE_BITS / 8))
#define I2S_NUM_CHANNELS (1)
#define I2S_SAMPLE_BITS (16)

 i2s_config_t i2s_config = {
      .mode = (i2s_mode_t)(I2S_MODE_MASTER | I2S_MODE_RX),
      .sample_rate = I2S_SAMPLE_RATE,
      .bits_per_sample = i2s_bits_per_sample_t(I2S_SAMPLE_BITS),
      .channel_format = I2S_CHANNEL_FMT_ONLY_LEFT,
      .communication_format = (i2s_comm_format_t)(I2S_COMM_FORMAT_I2S | I2S_COMM_FORMAT_I2S_MSB),
      .intr_alloc_flags = 0, // default interrupt priority
      .dma_buf_count = 64,
      .dma_buf_len = 1024,
      .use_apll = 1};

  esp_err_t result_install = i2s_driver_install(I2S_PORT_NUM, &i2s_config, 0, NULL);
  assert(result_install == ESP_OK);

  esp_err_t result_sample = i2s_set_sample_rates(I2S_PORT_NUM, I2S_SAMPLE_RATE); // set sample rates
  assert(result_sample == ESP_OK);

  // pin config
  const i2s_pin_config_t pin_config = {
      .bck_io_num = I2S_SCK,
      .ws_io_num = I2S_WS,
      .data_out_num = I2S_PIN_NO_CHANGE,
      .data_in_num = I2S_SD};

  esp_err_t result_pin = i2s_set_pin(I2S_PORT_NUM, &pin_config);
  assert(result_pin == ESP_OK);

  esp_err_t result_clk = i2s_set_clk(I2S_PORT_NUM, I2S_SAMPLE_RATE, I2S_BITS_PER_SAMPLE_16BIT, I2S_CHANNEL_MONO);
  assert(result_clk == ESP_OK);

  esp_err_t result_strt = i2s_start(I2S_PORT_NUM);
  assert(result_strt == ESP_OK);
```

and then in the `.ino` file when I want to get the sound I can do:
```cpp
char *audio_buff = (char *)calloc(I2S_READ_LEN, sizeof(char));
i2s_read(I2S_PORT_NUM, (void *)audio_buff, I2S_READ_LEN, &bytes_read, portMAX_DELAY);
```

But how would I do this in `devicescript`? and is this even possible?

**Comment by pelikhan** [comment]
Unfortunately, we do not have yet a i2s implementation (or microphone sample streaming either). To make this possible, one would need to implement a i2s service similar to i2c and surface it on the TypeScript side. @mmoskal pov?

**Reply by mmoskal** [reply]
I think it would be more like our SPI rather than I2C (DMA etc), but otherwise spot on!
